crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs

```diff
@@ -44,61 +44,58 @@ struct Actual {
     extra_stmts: Vec<Stmt>,
 }
 
-/// TODO: VisitMut
 #[fast_path(ShouldWork)]
-impl Fold for AsyncToGenerator {
-    noop_fold_type!();
+impl VisitMut for AsyncToGenerator {
+    noop_visit_mut_type!();
 
-    fn fold_module_items(&mut self, n: Vec<ModuleItem>) -> Vec<ModuleItem> {
-        self.fold_stmt_like(n)
+    fn visit_mut_module_items(&mut self, n: &mut Vec<ModuleItem>) {
+        self.visit_mut_stmt_like(n);
     }
 
-    fn fold_stmts(&mut self, n: Vec<Stmt>) -> Vec<Stmt> {
-        self.fold_stmt_like(n)
+    fn visit_mut_stmts(&mut self, n: &mut Vec<Stmt>) {
+        self.visit_mut_stmt_like(n);
     }
 }
 
 impl AsyncToGenerator {
-    fn fold_stmt_like<T>(&mut self, stmts: Vec<T>) -> Vec<T>
+    fn visit_mut_stmt_like<T>(&mut self, stmts: &mut Vec<T>)
     where
-        T: StmtLike + FoldWith<Actual>,
-        Vec<T>: FoldWith<Self>,
+        T: StmtLike + VisitMutWith<Actual>,
+        Vec<T>: VisitMutWith<Self>,
     {
-        let stmts = stmts.fold_children_with(self);
+        stmts.visit_mut_children_with(self);
 
-        let mut buf = Vec::with_capacity(stmts.len());
+        let stmts_len = stmts.len();
 
-        for stmt in stmts {
+        for idx in 0..stmts_len {
```

Comment:
You can use `stmts.take()` instead.

As we have to re-create a vector anyway, `take()` is fine.

Reply:
Vec is actually not recreated: previously there's new buf which appends existing stmt with new extra_stmt. But in here stmts are used as-is, only appends new extra_stmts. `take()` makes existing stmts drain.